source/io.c

```diff
@@ -421,3 +423,214 @@ napi_value aws_napi_io_socket_options_new(napi_env env, napi_callback_info info)
+             */
+            if (checked_offset < 0 || (uint64_t)checked_offset > total_bytes ||
+                (uint64_t)checked_offset < impl->bytes_read) {
+                result = AWS_IO_STREAM_INVALID_SEEK_POSITION;
+                goto failed;
+            }
+            final_offset = (uint64_t)checked_offset - impl->bytes_read;
+            break;
+        case AWS_SSB_END:
+            /* Offset must be negative, and must not be trying to go further back than the
+             * current length of the buffer, because those bytes have been purged
+             */
+            if (checked_offset > 0 || checked_offset == INT64_MIN || (uint64_t)(-checked_offset) > impl->buffer.len) {
+                result = AWS_IO_STREAM_INVALID_SEEK_POSITION;
+                goto failed;
+            }
+            final_offset = (uint64_t)impl->buffer.len - (uint64_t)(-checked_offset);
+            break;
+    }
+
+    AWS_ASSERT(final_offset <= SIZE_MAX);
+    size_t buf_offset = (size_t)final_offset;
+    AWS_ASSERT(buf_offset <= impl->buffer.len);
+
+    if (buf_offset == impl->buffer.len) {
+        impl->bytes_read += impl->buffer.len;
+        impl->buffer.len = 0;
+    } else if (buf_offset > 0) {
+        size_t new_len = impl->buffer.len - buf_offset;
+        memmove(impl->buffer.buffer, impl->buffer.buffer + buf_offset, new_len);
+        impl->buffer.len = new_len;
+    }
+
+failed:
+    aws_mutex_unlock(&impl->mutex);
+    return result;
+}
+
+static int s_input_stream_read(struct aws_input_stream *stream, struct aws_byte_buf *dest) {
+    struct aws_napi_input_stream_impl *impl = stream->impl;
+
+    size_t bytes_to_read = dest->capacity - dest->len;
+    if (bytes_to_read > impl->buffer.len) {
+        bytes_to_read = impl->buffer.len;
+    }
+
+    if (!aws_byte_buf_write(dest, impl->buffer.buffer, bytes_to_read)) {
+        return AWS_OP_ERR;
+    }
+
+    /* seek the stream past what's been read to advance the buffer/bytes_read */
+    aws_input_stream_seek(&impl->base, impl->bytes_read + bytes_to_read, AWS_SSB_BEGIN);
+    return AWS_OP_SUCCESS;
+}
+
+static int s_input_stream_get_status(struct aws_input_stream *stream, struct aws_stream_status *status) {
+    struct aws_napi_input_stream_impl *impl = stream->impl;
+    aws_mutex_lock(&impl->mutex);
+    status->is_end_of_stream = impl->eos;
+    aws_mutex_unlock(&impl->mutex);
+    status->is_valid = true;
+    return AWS_OP_SUCCESS;
+}
+
+static int s_input_stream_get_length(struct aws_input_stream *stream, int64_t *out_length) {
+    (void)stream;
+    (void)out_length;
+    return aws_raise_error(AWS_ERROR_UNIMPLEMENTED);
+}
+
+static void s_input_stream_clean_up(struct aws_input_stream *stream) {
+    struct aws_napi_input_stream_impl *impl = stream->impl;
+    struct aws_allocator *allocator = impl->buffer.allocator;
+    aws_mutex_clean_up(&impl->mutex);
+    aws_byte_buf_clean_up(&impl->buffer);
+    aws_mem_release(allocator, impl);
```

Comment:
The `aws_input_stream_clean_up()` "virtual" function calls `aws_mem_release(stream->allocator, stream);` on the stream.

IT SHOULDN'T. WE SHOULD FIX THIS NOW.

This was leading to a double-free in the Python code, and this code now has the same issue